Update Discussion API Load Tests

robrap opened the pull request:
#This branch is mostly based on Wajeeha's branch jia/dapi_MA-2476, which was used for the last round of performance tests, so this is our de facto performance tests.  However, there are commits from this branch and from Chris Lee's branch clee/dapi (upon which this is based), that never made it to master.

This branch also adds a commit that I required to fix some references that allowed this to run for me locally like some of the other tests, and includes some initial docs for how this was run.

I need help determining the state of these commits and what we want.

Reviewers:
- [x] @BenjiLee 
- [x] @wajeeha-khalid 

Post-review
- [x] squash commits

robrap commented on discussions_api/dapi.py:
```diff
@@ -20,18 +20,24 @@ class UnexpectedResponse(Exception):
     def __init__(self, *args, **kwargs):  # pylint: disable=super-on-old-class
         super(DiscussionsApiTasks, self).__init__(*args, **kwargs)
-        self.course_id = os.getenv('COURSE_ID')
+        #self.course_id = os.getenv('COURSE_ID')
```
Seems like we should used the environment variables and/or override with a config file, but not hard code these.  Agreed?

robrap replied:
@wajeeha-khalid @clee: There is a lot of commented out code throughout.  You can see the pep8 failures are mostly around this.  It would be great to either get PR comments or commits around your intentions around commented out code.  (This all may be from just one of you.)  Let me know what would be the fastest way to get to a PR that we think is really ready for review.

wajeeha-khalid replied:
The hard coded values were not supposed to be released to public repo; it was only for my personal use.
The hard coded values should be removed in favour of using environment variable.

robrap replied:
@dianakhuang: FYI: I am reworking this code.  Also, the CourseSeeder login is broken, and I am working on fixing that code as well.  In the meantime, I recommend you use this code as is and just manually create a course with the key from below: `course-v1:testX+C1+2016_C1`.

robrap commented on discussions_api/dapi.py:
```diff
@@ -128,29 +143,34 @@ def get_random_thread(self, page=None, page_size=None, prioritize_comments=False
+        #course_id = "course-v1:testX+LTC02+2016_LTC02"
+        #course_id = "course-v1:testX+LTC03+2016_LTC03"
+        #course_id = "course-v1:testX+LTC04+2016_LTC04"
+        course_id = "course-v1:testX+C1+2016_C1"
```
This should call out to same code above for handling course id.

wajeeha-khalid replied:
See my comment above.

robrap commented on discussions_api/dapi.py:
```diff
@@ -128,29 +143,34 @@ def get_random_thread(self, page=None, page_size=None, prioritize_comments=False
-            "course_id": urllib.quote_plus(self.course_id),
-            "following": random.choice(dapi_constants.FOLLOWING),
-            "view": random.choice(dapi_constants.VIEW),
+            #"course_id": urllib.quote_plus(self.course_id),
```
Do we need/want any of these commented out lines?

BenjiLee replied:
Initially when I was loadtesting the forums, a lot of the indexes were not set up. So I commented out some parameters that were sure to attempt a non-indexed query that we would not need. For example, we would never need to sort by asc if the desktop did not give us that option. I left the commented parameters in here because they are still valid parameters for the discussion api.

robrap commented on discussions_api/tasks/dapi_constants.py:
```diff
@@ -3,19 +3,20 @@
 FOLLOWING = [
     "true",
-    "false",
+    #"false",
```
@BenjiLee: I think these were in a PR that you did not want to go to master.  Can you help determine if you want some, all, or none of these changes?  Thank you.

robrap replied:
@dianakhuang: This file still needs clean-up.  I missed that.  I think it relates to the other comment as well about what query values we make available.

BenjiLee replied:
These are the constants that we pulled from at random. Unsure why "false" was taken out. I think it might be because there is no filter for posts that I am NOT following.

wajeeha-khalid replied:
The API only takes 'True' as value for following and yes there is no filter for otherwise.

robrap commented on discussions_api/tasks/task_helpers.py:
```diff
@@ -0,0 +1,196 @@
+    #course_id = "course_id=course-v1%3AtestX%2BLTC02%2B2016_LTC02"
+    #course_id = "course_id=course-v1%3AtestX%2BLTC03%2B2016_LTC03"
+    #course_id = "course_id=course-v1%3AtestX%2BLTC04%2B2016_LTC04"
+    course_id = "course_id=course-v1%3AtestX%2BC1%2B2016_C1"
```
We need to remember to update this line to handle the proper course id.  Is it fine just to go back to the first of these with `self.course_id`?

wajeeha-khalid replied:
Same as comment above. It should be reverted back to use self.course_id

robrap commented on discussions_api/tasks/task_helpers.py:
```diff
@@ -0,0 +1,196 @@
+    #  These parameters did seem to affect response times 09/29/16
+    page_size = "page_size={}".format(random.choice(dapi_constants.PAGE_SIZE))
+    #  These parameters did not seem to affect response times 09/29/16
+    # mark_as_read = "mark_as_read={}".format(random.choice(dapi_constants.MARK_AS_READ))
```
Again - delete or fix this?

robrap replied:
FYI: @dianakhuang: I forgot about this and didn't touch it.  Not sure if I should remove or enhance comment or do nothing.

BenjiLee commented on discussions_api/locustfile.py:
```diff
@@ -124,23 +124,33 @@ def on_start(self):
+        PatchThreadsTask: 92,
+        PatchCommentsTask: 83,
+        DeleteThreadsTask: 40,
+        DeleteCommentsTask: 300,
```
FYI: The previous numbers were based on forums behavior when I measured it in Nov 2015. 

more infor under testing strategy here https://openedx.atlassian.net/wiki/display/MA/Goals+and+setup

BenjiLee commented on discussions_api/seed_data/discussions_seeder.py:
```diff
@@ -64,9 +64,9 @@ def login_to_lms(self, email, password):
+    def seed_comments(self, course_id, posts, responses, child_comments):
         """
-        Creates 20 comments to a response for n posts
+        #Creates 20 comments to a response for n posts
```
Does this need a #?

robrap commented on discussions_api/discussions_api.py:
```diff
@@ -0,0 +1,79 @@
+# pylint: disable=no-member
```
@clee Do we need this file? Seems like a subset of dapi.py. Not sure if this has any changes we wish to keep?

robrap replied:
May have been @wajeeha-khalid. Not sure.

clee replied:
@robrap I'm not the clee you're looking for. :)

robrap replied:
Sorry about that @clee.  Hopefully this will be the last time I use your id if you want to stop following this thread. :)  Thank you.

@BenjiLee: The above was meant for you. :)

robrap replied:
FYI: I'm deleting this file.

BenjiLee commented on discussions_api/dapi.py:
```diff
@@ -128,29 +138,31 @@ def get_random_thread(self, page=None, page_size=None, prioritize_comments=False
         """
+
+        # The following are additional query_args that we should consider
+        # load testing against:
```
Maybe more details:

`The following are additional query_args that we should consider load testing against. Adding a query that is not indexed may throw off the load test results. For example, if there is an index for order_by=asc but no index for order_by=desc, the locust results may be very hard to read.`

robrap replied:
@BenjiLee: As a general rule, it feels like we either shouldn't allow a particular query param or we should be sure it performs, right? It seems odd having code that someone could run that may not perform.  Maybe I'm missing something.

BenjiLee replied:
This was also a conversation when the discussion API was first created. Why allow someone to call asc if it's not index. I guess my goal here was that the discussion API allows for possible queries that are not index, and I wanted to make note of it.

wajeeha-khalid replied:
For larger response, `order_by=desc` sometimes crashes the response (supposedly because its not indexed) and thats is why order_direction was commented out.

robrap replied:
I'm going to temporarily drop this code until questions get answered regarding these parameters in this new JIRA ticket: https://openedx.atlassian.net/browse/TNL-4923

robrap replied:
I also created a ticket to add test for "following" filter, which seems to have issues. https://openedx.atlassian.net/browse/TNL-4929

BenjiLee commented on discussions_api/locustfile.py:
```diff
@@ -124,23 +125,26 @@ def on_start(self):
         params = {
             "course_id": self.course_id,
-            "staff": "true",
-            "roles": ["Administrator"]
```
Any particular reason this was removed?

robrap replied:
I'm happy to add it back if it belongs @BenjiLee.
@wajeeha-khalid: Any reason this got dropped, or was this a mistake?

wajeeha-khalid replied:
It was not working for me @BenjiLee; if I add role it gives me error that no auth endpoint exists for administrator.

BenjiLee replied:
Fine with dropping it then. If in the future, a staff user is unable to make moderator requests like edit or delete, take a look at roles.

robrap replied:
I added this back in and it seems to be running.  I'll leave it in until it causes an issue.

BenjiLee commented on discussions_api/seed_data/seed_data.py:
```diff
@@ -37,15 +38,62 @@ def setup_lms(args):
+        seeder: The DiscussionSeeder.
+        file_name: The filename to be written to.
+        course_id: The course containing the threads.
+        thread_id_list: The thread list, which will be retrieved from the course if None is supplied.
```
nits: Adding the argument type would be helpful in this doc. Not immediately sure if thread_id_list is a list or a string that identifies a list of threads.

robrap replied:
On it's way.

BenjiLee commented on discussions_api/seed_data/seed_data.py:
```diff
@@ -115,79 +155,56 @@ def setup_flow_test(args):
+    args.threads = 6
+    args.responses = 6
+    args.comments = 6
+    # TODO: This is not saving the earlier threads to the file. Does this need to be enhanced?
```
By file do you mean the course we are creating the threads on? 

Also, not sure what this TODO means.

robrap replied:
I meant the SEED_DATA file that is written out with the thread ids to be used in the locust test.  I decided to delete this function for now so we can just merge something reasonably clean, and I added the following task for us to re-implement something that sets up a reasonable test course:
https://openedx.atlassian.net/browse/TNL-4930

BenjiLee commented on discussions_api/tasks/dapi_constants.py:
```diff
@@ -3,19 +3,20 @@
 VIEW = [
     "unread",
-    "unanswered",
+    # "unanswered",
```
At the time there was no index for this and was commented out? There might be a ticket somewhere to add the unanswered query. I'll try to dig it up/

robrap replied:
Ok.  I will add this back in.

BenjiLee commented on discussions_api/tasks/dapi_constants.py:
```diff
@@ -32,6 +33,20 @@
+#    "10",
+#    "20"
+#]
+#PAGE_SIZE = [
```
Sometimes page size can affect the response times. I guess it would depend on what someone is trying to load test. 

Should we remove these? and just add a comment regarding variable page size?

robrap replied:
I've created a ticket around this: https://openedx.atlassian.net/browse/TNL-4931

What I want to do is get a working codebase merged that we will then immediately iterate upon for our specific needs.

wajeeha-khalid commented on discussions_api/dapi.py:
```diff
@@ -128,29 +138,31 @@ def get_random_thread(self, page=None, page_size=None, prioritize_comments=False
-            "course_id": urllib.quote_plus(self.course_id),
-            "following": random.choice(dapi_constants.FOLLOWING),
-            "view": random.choice(dapi_constants.VIEW),
+            "course_id": self.course_id,
```
When using env param for course; you would want to encode it as line# 132

robrap replied:
@wajeeha-khalid: I don't think I do because we would all be grabbing the course id out of the url and it already works as is.  When I call `urllib.quote_plus()` on something like `course-v1:dapi+FlowTest+test_e779d`, it doesn't work.

wajeeha-khalid commented on discussions_api/seed_data/discussions_seeder.py:
```diff
@@ -1,106 +1,71 @@
+        Returns:
+            The list of thread ids created.
 
-    def seed_comments(self, course_id, posts):
```
This method could also be reverted back to @BenjiLee 's branch to create random number of comments and responses per thread which is a more ideal scenario.
I had refactored it for my personal use to input same no. of comments and responses to seed for all threads.

robrap replied:
This is a reminder to myself of where I left off....

robrap replied:
@wajeeha-khalid: I'm going to leave this as-is for now.  It seems reasonable to have parameters and then let the caller decide, or have another wrapping function.

wajeeha-khalid commented on discussions_api/seed_data/seed_data.py:
```diff
@@ -60,52 +108,44 @@ def create_threads(args, course_id=None, seeder=None, save_threads=True):
+    responses = int(args.responses or raw_input("How many responses for thread "))
+    child_comments = int(args.comments or raw_input("How many comments for each response "))
+    thread_list = seeder.seed_comments(course_id=course_id, posts=posts, responses=responses, child_comments=child_comments)
+    file_name = args.action + str(responses * child_comments)
```
Again (in context of my last comment to seed comments), this could be reverted back to @BenjiLee 's branch.

wajeeha-khalid commented on discussions_api/seed_data/seed_data.py:
```diff
@@ -60,52 +108,44 @@ def create_threads(args, course_id=None, seeder=None, save_threads=True):
+    seeder = seeder if seeder else setup_discussion_seeder(args)
+    posts = int(args.threads or raw_input("How many threads "))
+    responses = int(args.responses or raw_input("How many responses for thread "))
+    child_comments = int(args.comments or raw_input("How many comments for each response "))
```
Again (in context of my last comment to seed comments), this could be reverted back to @BenjiLee 's branch.

robrap replied:
@wajeeha-khalid: I'm going to leave this as-is for now.  And then I may adjust as I start to use this.

wajeeha-khalid commented on discussions_api/seed_data/seed_data.py:
```diff
@@ -115,79 +155,56 @@ def setup_flow_test(args):
+        ("GetCommentsTasks", create_comments),
+        ("GetThreadsTask", create_threads),
+        ("PatchThreadsTask", create_threads),
+        ("PostThreadsTask", return_course),
```
Do we really need all these types of tasks while the purpose of all these is same as to create some initial forums data?
May be only two are enough:
create_threads
create_threads_with_comments

BenjiLee replied:
This can be redone. I have no strong feelings to keeping this. I think when I first created these, I was thinking of having known data sets that went with known locust tasks.

robrap replied:
I'm going to leave as-is for now until I start really putting this to use.

wajeeha-khalid commented:
There are two files containing almost same helper functions `dapi.py` and `tasks/task_helpers.py`. Would be great if we combine them in single file. As far as I remember, I tried to import all methods from `dapi.py`. So, may be we can remove `tasks/task_helpers.py` and rename `dapi.py` to a more readable name.

robrap commented:
FYI: @ormsbee @tobz

BenjiLee commented on discussions_api/tasks/dapi_tasks.py:
```diff
@@ -84,11 +96,40 @@ def get_comment_list(self):
         GET_comment_list
         """
-        thread = self.get_random_thread(page=1, page_size=100, prioritize_comments=True)
+        #choice = random.choice(self.thread_id_list)
```
@wajeeha-khalid Unsure what the commented code is.

wajeeha-khalid replied:
It can be removed.

BenjiLee commented on discussions_api/tasks/dapi_tasks.py:
```diff
@@ -112,15 +153,21 @@ def patch_edit_thread(self):
 
         Calls made:
-        GET_thread_list
+        POST_thread
```
@wajeeha-khalid A bit confused on the change of this documentation (and the ones below it).

wajeeha-khalid replied:
Patch has some restrictions based on access of user on a thread and it resulted in many unsuccessful patch requests if we pick a random thread and make patch on it.
Hence I had updated to create a thread and make patch on it with the owner rights.

Does it make sense?

wajeeha-khalid replied:
It is referring to the commented code. We can remove commented code and revert back the doc string.

BenjiLee commented on discussions_api/tasks/dapi_tasks.py:
```diff
@@ -196,7 +255,7 @@ class PatchCommentsTask(DiscussionsApiTasks):
 
     def _get_comment_id(self):
-        thread = self.get_random_thread(page=1, page_size=100, prioritize_comments=True)
+        thread = self.get_random_thread(page=1, page_size=10, prioritize_comments=True)
```
If reducing the page_size (for performance reason) here, than we should vary the page we select. Otherwise, we would be picking a comment from the same 10 threads. We can also not set `page`. In get_random_thread, if `page` is not set, a random page will be selected.

wajeeha-khalid replied:
+1

BenjiLee commented on discussions_api/tasks/dapi_tasks.py:
```diff
@@ -316,15 +401,16 @@ def post_comment(self):
-        POST_comment_comment
         """
-        thread = self.get_random_thread(page=1, page_size=100)
+        thread = self.get_random_thread(page=1, page_size=10)
```
Same comment about random threads here.

robrap replied:
Adding some randomization.

BenjiLee commented:
It seems like the sequence of calls for each call (get thread, delete thread etc...) of discussion API calls have changed in this dapi_tasks.py. The original ratios in the locust tasks were based on these. Maybe thats why the ratios have changed?

BenjiLee commented on discussions_api/tasks/task_helpers.py:
```diff
@@ -0,0 +1,191 @@
+    """
+    #  These parameters did seem to affect response times 09/29/16
+    page_size = "page_size={}".format(random.choice(dapi_constants.PAGE_SIZE))
+    #  These parameters did not seem to affect response times 09/29/16
```
I would remove these comments. Novice me left behind bad comments that could be misleading/stale eventually.

robrap replied:
@BenjiLee: This entire file was deleted.  At some point it looks like this code was moved to dapi.py, and I didn't see this code in use.

BenjiLee commented on helpers/auto_auth_tasks.py:
```diff
@@ -45,6 +45,7 @@ def auto_auth(self, verify_ssl=True, params=None, hostname=''):
             params=params or {},
             verify=verify_ssl
         )
+
```
This added line is blasphemous

BenjiLee commented:
Done with pass.

robrap commented:
@clee @wajeeha-khalid: Thanks for all the notes.  I'll be working through this again next week.

FYI: @clee @wajeeha-khalid @ormsbee @dianakhuang: I want to accomplish to things:
1) Merge something to master that isn't broken, and
2) Get to the types of tests I need to run to find and fix known issues as soon as possible.

In order to do this, my current approach for this PR is simply to get to a generally working and clean code base and move any issues that are lower priority that number 2 above to backlog JIRA tickets.  This way we can merge quicker, get to the testing we need quicker, and come back to other issues at a later time.

Let me know if anyone has an issue with this approach.  Thanks.

robrap commented:
Another note to self: run pep8 on other directories before push to clean up Travis failures.

BenjiLee commented:
@robrap SGTM. Let me know when you are ready for another pass

BenjiLee commented:
Just a heads up, the load tests and data seeder were NOT designed with testing discussion topics in mind.

robrap commented:
DYI: @BenjiLee @wajeeha-khalid @dianakhuang @ormsbee: I'm done with my updates based on code review comments.
Again, I'm just trying to get master in a reasonably stable place so I can start the next round of changes for my own seeding and testing needs.

wajeeha-khalid commented on discussions_api/seed_data/seed_data.py:
```diff
@@ -60,134 +109,84 @@ def create_threads(args, course_id=None, seeder=None, save_threads=True):
+        ("PatchThreadsTask", create_threads),
+        ("PostThreadsTask", return_course),
+        ("SaveThreads", save_course_thread_list_to_file),
+    ])
```
@robrap  In my opinion; these tasks should be generalised to seed data tasks rather than creating one task per discussion endpoint.

wajeeha-khalid commented:
@robrap Just one comment about seed tasks which could also be entertained in a separate PR; else looks good to be merged

robrap commented:
@wajeeha-khalid: I simplified and renamed the seed actions based on your comment.  I also found one minor big left in one of the actions.
@BenjiLee: FYI ^^^

@wajeeha-khalid @BenjiLee: Let me know if there is anything else blocking thumbs.  Thanks.

BenjiLee commented on discussions_api/seed_data/seed_data.py:
```diff
@@ -60,134 +110,88 @@ def create_threads(args, course_id=None, seeder=None, save_threads=True):
-    course_id = course_id if course_id else setup_studio(args)
-    seeder = seeder if seeder else setup_lms(args)
-    posts = int(args.number or raw_input("How many threads in multiples of 10?"))
+    course_id = get_course_id(args, course_id)
```
IS the docstring for this file still accurate?

robrap replied:
I think so.  Any reason to believe it is not?

BenjiLee replied:
Just double checking.

BenjiLee commented on discussions_api/tasks/dapi_constants.py:
```diff
@@ -28,9 +25,16 @@
+    "20",
+]
+PAGE_SIZE = [
+    "1",
```
I think page_sizes can create unexpected resulted. For defaults, it might be better to just have 1 size (default on web is 20).

BenjiLee commented on discussions_api/tasks/dapi_tasks.py:
```diff
@@ -316,15 +349,17 @@ def post_comment(self):
         Calls made:
         GET_thread_list
         POST_comment_response
-        POST_comment_comment
```
This should be added back in

BenjiLee commented on discussions_api/tasks/dapi_tasks.py:
```diff
@@ -340,22 +375,17 @@ def delete_thread(self):
 
-        A thread will be created with POST, then randomly selected with GET,
-        and then DELETED.
+        A thread will be created with POST and then DELETED.
```
Note: By changing the sequence of calls here, the task ratio needs to adjusted

robrap replied:
Once again, I am punting and have created a JIRA story:
https://openedx.atlassian.net/browse/TNL-4979

BenjiLee commented on discussions_api/tasks/dapi_tasks.py:
```diff
@@ -371,54 +401,50 @@ def delete_response(self):
-        POST_comment
+        GET_response_list
+        POST_response_comment
         DELETE_comment
```
might as well call this `DELETE_response_comment"

BenjiLee commented:
Done with pass. 

LGTM 👍 

The biggest thing that needs to be updated is the ratio of forum calls. Since some of the sequence of calls have been modified, the ratio will change. 

Also, is it sound to create a post and then delete said post?

robrap commented:
@BenjiLee @wajeeha-khalid: I'll be going through your comments.  In the meantime, I did remove the "view" parameter because it was causing some issues and I added a story to deal with this: https://openedx.atlassian.net/browse/TNL-4978

Thanks.

robrap commented:
@BenjiLee: I know you already gave thumbs, but I pushed fixes for your final comments (and added one new ticket).

BenjiLee commented:
👍

robrap commented:
@tobz: Let me know if you think it is ok to merge this.  Mostly I was trying to get this to a state where it was runnable.  It is in fact runnable.  However, I fixed some things and removed/punted on other things and moved them to separate JIRA tickets.  Do you think this is good to go?  Master is broken until this gets merged.

FYI: @pwnage101

wajeeha-khalid commented:
👍

pwnage101 commented on discussions_api/seed_data/course_seeder.py:
```diff
@@ -101,27 +49,26 @@ def import_tarfile(self, course_id, tarfile):
+        headers.pop("Content-Type")
 
-        with open(tarfile, 'rb') as upload:
+        with open(os.path.dirname(__file__) + '/' + tarfile, 'rb') as upload:
```
for path manipulation safety, could you change this to use os.path.join() ?

pwnage101 commented on discussions_api/seed_data/discussions_seeder.py:
```diff
@@ -168,46 +132,54 @@ def create_comment(self, thread_id, comment_is_child=None):
-                print "{}: Could not get next in response. Trying again in 5 seconds".format(response.status_code)
-                time.sleep(5)
-                response = self.sess.get(next_page_url)
+        if not thread_id_list:
```
With no else block here, `thread_id_list` essentially gets passed from input to output directly.  Shouldn't this variable, if given, at least be used as a filter over thread ids returned by the API?  e.g.:

`returned_thread_ids = set(thread_id_list).intersect(fetched_thread_ids)`

robrap replied:
I don't think this is necessary for how this code gets used.  It would probably be cleaner if this was split into two different functions though:
1. Write the thread list to a file.
2. Read all threads from a course into the list.
I don't think the filter is needed because mainly this code is what is creating all of these threads and writing it to a file, so it seems unnecessary to go back and check if these newly created threads are really part of the course.

I might just add a TODO comment.  

You missed an earlier comment that I had written.  I was trying to put in some minimal amount of work to get this runnable and create additional JIRA tasks for improvements, but I am also trying to close this as quickly as possible to get master in a working condition and to move on to finding some known performance issues, which is where I need to be concentrating my time.

So, we'll need to be careful about what we think is absolutely necessary to get this merged back into master, which is actively broken.

Thanks.

pwnage101 replied:
Okay, sounds good!  Please at least add a TODO if that's easier than creating a JIRA task, it looks like a future bug to me.

pwnage101 commented:
oops, okay i'm done with my "pass" :P

robrap commented:
No worries @pwnage101.  I appreciate the comments.  I just wanted you to understand the context.  Thanks!

pwnage101 commented:
:+1:
